homeassistant/components/recorder/migration.py

```diff
@@ -652,6 +652,18 @@ def _apply_update(instance, new_version, old_version):  # noqa: C901
     elif new_version == 27:
         _add_columns(instance, "events", [f"data_id {big_int}"])
         _create_index(instance, "events", "ix_events_data_id")
+    elif new_version == 28:
+        _add_columns(instance, "events", ["origin_idx INTEGER"])
+        # We never use the user_id or parent_id index
+        _drop_index(instance, "events", "ix_events_context_user_id")
+        _drop_index(instance, "events", "ix_events_context_parent_id")
+        _add_columns(instance, "states", ["origin_idx INTEGER"])
+        _add_columns(instance, "states", ["context_id VARCHAR(36)"])
+        _add_columns(instance, "states", ["context_user_id VARCHAR(36)"])
+        _add_columns(instance, "states", ["context_parent_id VARCHAR(36)"])
```

Comment:
Why not pass multiple columns to `_add_columns`? Certain database engines can optimize for this.

Reply:
Also, would it make sense to back-fill this information so we can simplify the queries we use right away?

Reply:
We don't need it since we get it from the join from old rows.  We can take the joins out after a few months

Reply:
Will switch it to pass multiple to `_add_columns`

Reply:
333e1502da61da7023e4d2f9cf9c1c12bad96f34

Reply:
> Also, would it make sense to back-fill this information so we can simplify the queries we use right away?

I think I can reduce it without having to backfill.. 🤞  https://github.com/bdraco/home-assistant/commit/8c767b88996811d9ed63a39745ba395a4c58bfc4

Testing it now

Reply:
The query looks fast since we also use a start date with the automation logbook context lookup

```
	SELECT anon_1.event_type AS anon_1_event_type, anon_1.event_data AS anon_1_event_data, anon_1.time_fired AS anon_1_time_fired, anon_1.context_id AS anon_1_context_id, anon_1.context_user_id AS anon_1_context_user_id, anon_1.context_parent_id AS anon_1_context_parent_id, anon_1.shared_data AS anon_1_shared_data, anon_1.state AS anon_1_state, anon_1.entity_id AS anon_1_entity_id, anon_1.attributes AS anon_1_attributes, anon_1.shared_attrs AS anon_1_shared_attrs 
FROM (SELECT events.event_type AS event_type, events.event_data AS event_data, events.time_fired AS time_fired, events.context_id AS context_id, events.context_user_id AS context_user_id, events.context_parent_id AS context_parent_id, event_data.shared_data AS shared_data, NULL AS state, NULL AS entity_id, NULL AS attributes, NULL AS shared_attrs 
FROM events LEFT OUTER JOIN event_data ON events.data_id = event_data.data_id 
WHERE events.time_fired > '2022-05-02 05:43:09.002362' AND events.time_fired < '2022-05-03 05:43:09.002362' AND events.event_type IN ('logbook_entry', 'call_service', 'homeassistant_start', 'homeassistant_stop', 'automation_triggered', 'script_started', 'homekit_state_change') AND events.context_id = '01G21MPGEA596Y7H8QCDCHF98X' UNION ALL SELECT 'state_changed' AS event_type, NULL AS event_data, states.last_changed AS time_fired, states.context_id AS context_id, states.context_user_id AS context_user_id, states.context_parent_id AS context_parent_id, NULL AS shared_data, states.state AS state, states.entity_id AS entity_id, states.attributes AS attributes, state_attributes.shared_attrs AS shared_attrs 
FROM states LEFT OUTER JOIN states AS old_state ON states.old_state_id = old_state.state_id LEFT OUTER JOIN state_attributes ON states.attributes_id = state_attributes.attributes_id LEFT OUTER JOIN events ON states.event_id = events.event_id 
WHERE old_state.state_id IS NOT NULL AND states.state != old_state.state AND states.state IS NOT NULL AND ((states.entity_id NOT LIKE 'sensor.%' AND states.entity_id NOT LIKE 'proximity.%') OR ((states.entity_id LIKE 'sensor.%' OR states.entity_id LIKE 'proximity.%') AND (state_attributes.shared_attrs NOT LIKE '%\"unit_of_measurement\":%' OR states.attributes NOT LIKE '%\"unit_of_measurement\":%'))) AND states.last_updated > '2022-05-02 05:43:09.002362' AND states.last_updated < '2022-05-03 05:43:09.002362' AND states.last_updated = states.last_changed AND (states.context_id = '01G21MPGEA596Y7H8QCDCHF98X' OR events.context_id = '01G21MPGEA596Y7H8QCDCHF98X')) AS anon_1 ORDER BY anon_1.time_fired
```


```
+------+-------------+------------------+------------+-----------------------------------------------------------------------------------------------------+------------------------------------------------------+---------+------------------------------------+--------+--------------------------------------------------------+
| id   | select_type | table            | type       | possible_keys                                                                                       | key                                                  | key_len | ref                                | rows   | Extra                                                  |
+------+-------------+------------------+------------+-----------------------------------------------------------------------------------------------------+------------------------------------------------------+---------+------------------------------------+--------+--------------------------------------------------------+
|    1 | PRIMARY     | <derived2>       | ALL        | NULL                                                                                                | NULL                                                 | NULL    | NULL                               | 16     | Using filesort                                         |
|    2 | DERIVED     | events           | ref|filter | ix_events_time_fired,ix_events_event_type_time_fired,ix_events_context_id                           | ix_events_context_id|ix_events_event_type_time_fired | 147|268 | const                              | 2 (0%) | Using index condition; Using where; Using rowid filter |
|    2 | DERIVED     | event_data       | eq_ref     | PRIMARY                                                                                             | PRIMARY                                              | 4       | homeassistant.events.data_id       | 1      | Using where                                            |
|    3 | UNION       | states           | range      | ix_states_old_state_id,ix_states_last_updated,ix_states_entity_id_last_updated,ix_states_context_id | ix_states_last_updated                               | 9       | NULL                               | 29     | Using where                                            |
|    3 | UNION       | state_attributes | eq_ref     | PRIMARY                                                                                             | PRIMARY                                              | 4       | homeassistant.states.attributes_id | 1      | Using where                                            |
|    3 | UNION       | events           | eq_ref     | PRIMARY                                                                                             | PRIMARY                                              | 4       | homeassistant.states.event_id      | 1      | Using where                                            |
|    3 | UNION       | old_state        | eq_ref     | PRIMARY                                                                                             | PRIMARY                                              | 4       | homeassistant.states.old_state_id  | 1      | Using where                                            |
+------+-------------+------------------+------------+-----------------------------------------------------------------------------------------------------+------------------------------------------------------+---------+------------------------------------+--------+--------------------------------------------------------+
```

Reply:
So the worst case would be if they wanted to find a context id from an old automation trace from a few days ago, before the format changed, we would have match all the context ids for a one day period.  That took about 8s on the test db I had with 2 million state rows.  I think this isn't too much of a concern.

Reply:
I went ahead and simplified the queries

Reply:
I just realized I can simplify this a bit more since we don't need multiple event query selects anymore.  I'll do that in a followup though

Reply:
Slept on it and changed my mind about it and opened https://github.com/home-assistant/core/pull/71201

I'm probably being overly paranoid but I should remember users always find a way to use it in a way you don't expect.  I isolated the legacy query into its own function so its much easier to strip out later